### PR TITLE
Properly initialize reactive Pool beans

### PR DESCRIPTION
**Pull request by geoand**
We introduce an injection point to the
synthetic bean creation process thus
giving Arc all the declarative context
it needs in order to ensure proper bean
creation order

Relates to https://github.com/quarkiverse/quarkus-vault/pull/167

### CODE REVIEW

File: extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/VertxPoolBuildItem.java [hunk 1]
```diff
@@ -1,37 +0,0 @@
...
-import io.quarkus.runtime.RuntimeValue;
-import io.vertx.sqlclient.Pool;
-
-/**
- * A build item for Vertx {@link Pool}s.
- * <p>
- * If you inject this build item when recording runtime init template calls, you are guaranteed the Pool configuration
- * has been injected and Pools can be created.
- */
-public final class VertxPoolBuildItem extends MultiBuildItem {
```
Comment (geoand) [inline extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/VertxPoolBuildItem.java:13]: This is not used anywhere in a our code and there is no good reason why anyone else would use it either

Comment (geoand) [reply extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/VertxPoolBuildItem.java:13]: We could potentially keep the build item around just throw an exception if the getter methods are called.

File: extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/VertxPoolBuildItem.java [hunk 2]
```diff
@@ -12,26 +12,23 @@
  */
 public final class VertxPoolBuildItem extends MultiBuildItem {
```
Comment (gsmet) [inline extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/VertxPoolBuildItem.java:13]: Should we mark it as deprecated and `forRemoval` then?

**Review by gsmet — COMMENTED** [review]
Look good overall, I added a small comment.

File: extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/VertxPoolBuildItem.java (hunk 2 above)
Comment (geoand) [reply extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/VertxPoolBuildItem.java:13]: Good point!

Comment (geoand) [reply extensions/reactive-datasource/deployment/src/main/java/io/quarkus/reactive/datasource/deployment/VertxPoolBuildItem.java:14]: Done

**Review by yrodiere — APPROVED** [review]

**Review by tsegismont — APPROVED** [review]
LGTM, thank you @geoand 

Just one question about a code comment

File: extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolRecorder.java [hunk 3]
```diff
@@ -43,42 +45,59 @@
...
-                dataSourcesRuntimeConfig.dataSources().get(dataSourceName),
-                dataSourcesReactiveRuntimeConfig.getDataSourceReactiveRuntimeConfig(dataSourceName),
-                dataSourcesReactivePostgreSQLConfig.dataSources().get(dataSourceName).reactive().postgresql());
-
-        shutdown.addShutdownTask(pgPool::close);
-        return new RuntimeValue<>(pgPool);
+        return new Function<>() {
+            @Override
+            public PgPool apply(SyntheticCreationalContext<PgPool> context) {
+                // Only create the OtlpGrpcSpanExporter if an endpoint was set in runtime config and was properly validated at startup
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolRecorder.java:61]: Oops! Thanks for spotting, I'll remove it

**Comment by vsevel** [comment]
nice change. is the goal to backport this on `3.4.0`, and finalize https://github.com/quarkiverse/quarkus-vault/pull/167?

**Comment by geoand** [comment]
I am not sure we can backport because this breaks the use of some build items (granted they are unlikely to be used)

**Comment by vsevel** [comment]
hmm. so what do we do for vault in `3.4` ?
I am confused. we actually pulled your branch https://github.com/quarkiverse/quarkus-vault/pull/167 and used it against `CR1`, and it worked on your sample app.
is this change appropriate in `3.4` (or not)?
if not, what can we do in vault for `3.4`? we can't leave the `VertxRecorder`, can we?

**Comment by geoand** [comment]
For 3.4 you should use https://github.com/quarkiverse/quarkus-vault/pull/167 which you said seems to work.

This change here in just to sure that we don't run into issues that lead to the use of the VertxRecorder in the first place.

**Comment by vsevel** [comment]
well. it works in the use case we tested.
if we go ahead then we are probably going to re-ignite issue https://github.com/quarkiverse/quarkus-vault/issues/78, which lead you to do do https://github.com/quarkiverse/quarkus-vault/pull/109 as @kdubb was reporting
it seems that either way we can't have a solution working at 100%.
if we don't do anything we have https://github.com/quarkiverse/quarkus-vault/issues/165 (and nothing works)
and if we do https://github.com/quarkiverse/quarkus-vault/pull/167 we are going to have https://github.com/quarkiverse/quarkus-vault/issues/78

if correct, I think we need to go ahead with https://github.com/quarkiverse/quarkus-vault/pull/167, release vault `3.2`, and report in the migration guide that this version does not work with reactive datasources, which will be fixed in `3.5`
any other (better) option?

**Comment by geoand** [comment]
> if we go ahead then we are probably going to re-ignite issue https://github.com/quarkiverse/quarkus-vault/issues/78, which lead you to do do https://github.com/quarkiverse/quarkus-vault/pull/109 as @kdubb was reporting

any way to verify this?

**Comment by vsevel** [comment]
> any way to verify this?

@vmargerin can you check if we still have the issue with `CR1` and https://github.com/quarkiverse/quarkus-vault/pull/167 using https://github.com/quarkiverse/quarkus-vault/issues/78#issuecomment-1413928199?

**Comment by geoand** [comment]
Thanks @vsevel.

In that case, we indeed might want to consider this for 3.4... @cescoffier @gsmet WDYT? This is technically breaking as it changes a build item, but it is **highly** unlikely these build items are used by any external extension.

**Comment by geoand** [comment]
The backport itself should be easy

**Pull request merged**